### ISSUE TITLE
Cannot read properties of undefined (reading 'buffers') - Last version

### ISSUE THREAD

**Issue by MarianoFacundoArch** [redis/node-redis#1734]
I get this kill randomly,
can't find reason. This happens on version 4.

0|index  | TypeError: Cannot read properties of undefined (reading 'buffers')
0|index  |     at Function._RedisCommandsQueue_emitPubSubMessage (/opt/build/node_modules/@node-redis/client/dist/lib/client/commands-queue.js:215:38)
0|index  |     at JavascriptRedisParser.returnReply (/opt/build/node_modules/@node-redis/client/dist/lib/client/commands-queue.js:43:123)
0|index  |     at JavascriptRedisParser.execute (/opt/build/node_modules/redis-parser/lib/parser.js:544:14)
0|index  |     at RedisCommandsQueue.parseResponse (/opt/build/node_modules/@node-redis/client/dist/lib/client/commands-queue.js:190:71)
0|index  |     at RedisSocket.<anonymous> (/opt/build/node_modules/@node-redis/client/dist/lib/client/index.js:323:83)
0|index  |     at RedisSocket.emit (node:events:390:28)
0|index  |     at Socket.<anonymous> (/opt/build/node_modules/@node-redis/client/dist/lib/client/socket.js:183:44)
0|index  |     at Socket.emit (node:events:390:28)
0|index  |     at addChunk (node:internal/streams/readable:315:12)
0|index  |     at readableAddChunk (node:internal/streams/readable:289:9)
0|index  |     at Socket.Readable.push (node:internal/streams/readable:228:10)
0|index  |     at TCP.onStreamRead (node:internal/stream_base_commons:199:23)
0|index  | TypeError: Cannot read properties of undefined (reading 'buffers')
0|index  |     at Function._RedisCommandsQueue_emitPubSubMessage (/opt/build/node_modules/@node-redis/client/dist/lib/client/commands-queue.js:215:38)
0|index  |     at JavascriptRedisParser.returnReply (/opt/build/node_modules/@node-redis/client/dist/lib/client/commands-queue.js:43:123)
0|index  |     at JavascriptRedisParser.execute (/opt/build/node_modules/redis-parser/lib/parser.js:544:14)
0|index  |     at RedisCommandsQueue.parseResponse (/opt/build/node_modules/@node-redis/client/dist/lib/client/commands-queue.js:190:71)
0|index  |     at RedisSocket.<anonymous> (/opt/build/node_modules/@node-redis/client/dist/lib/client/index.js:323:83)
0|index  |     at RedisSocket.emit (node:events:390:28)
0|index  |     at Socket.<anonymous> (/opt/build/node_modules/@node-redis/client/dist/lib/client/socket.js:183:44)
0|index  |     at Socket.emit (node:events:390:28)
0|index  |     at addChunk (node:internal/streams/readable:315:12)
0|index  |     at readableAddChunk (node:internal/streams/readable:289:9)
0|index  |     at Socket.Readable.push (node:internal/streams/readable:228:10)
0|index  |     at TCP.onStreamRead (node:internal/stream_base_commons:199:23)

**Comment by leibale** [comment]
It seems like it has something to do with PubSub, can you please share that part from your code?

**Comment by Rmaan** [comment]
I also got the same error. I can't create a reproduce right now (it doesn't happen always) but the gist is: I subscribe to channel then I unsubscribe and if I get a message on that channel at the same moment I will run into this error.

I looked a bit at code and it seems we have a race condition:
Error happens on [this line](https://github.com/redis/node-redis/blob/0b2a8d7c3c75f504ecde80bc0e2c64e57a2bff0f/packages/client/lib/client/commands-queue.ts#L77) because assumption was listener map always contains the key but it's not the case.
When I call unsubscribe, [this line](https://github.com/redis/node-redis/blob/0b2a8d7c3c75f504ecde80bc0e2c64e57a2bff0f/packages/client/lib/client/commands-queue.ts#L242) will be executed *before* sending the command to Redis, and if a message arrive after execution of this line and before sending the message then error happens.
